### PR TITLE
Implemented zoom functionality

### PR DESCRIPTION
**Pull request by WyattTheSkid**
An OptiFine replacement needs to replace the zoom feature too.

### CODE REVIEW

**Comment by huihiuhuai** [comment]
can you also fix [this](https://github.com/Nolij/Zume/issues/26)?

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @SubscribeEvent
+    public void ZoomyZoomZoom(FOVUpdateEvent e) {
+        if (zoomKey.getIsKeyPressed() || zoomKey.isPressed()) {
+            zoom = true;
+            if (zoom) {
+                // Adjust zoom level with mouse wheel while holding zoom key
+                int dWheel = Mouse.getDWheel();
+                if (dWheel != 0) {
```
Comment (eigenraven) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:35]: This entire if can be replaced with `zoomamt -= Integer.signum(dWheel);`

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import cpw.mods.fml.common.Mod;
+import cpw.mods.fml.common.event.FMLInitializationEvent;
+import cpw.mods.fml.common.eventhandler.SubscribeEvent;
+import net.minecraft.client.settings.KeyBinding;
+import net.minecraftforge.client.event.FOVUpdateEvent;
+import net.minecraftforge.common.MinecraftForge;
+import org.lwjgl.input.Keyboard;
+import org.lwjgl.input.Mouse;
+
+@Mod(modid = "skidzoom", name = "SkidZoom", version = "1.0")
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:13]: please don't create another mod, just hook it up to angelica

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+public class SkidZoom {
+
+    public KeyBinding zoomKey;
+    private static boolean zoom;
+    private int zoomamt = -7;
+
+    @Mod.EventHandler
+    public void init(FMLInitializationEvent event) {
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:21]: only register stuff on the client

File: src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java [hunk 4]
```diff
@@ -530,7 +530,13 @@ public enum Mixins {
         .addMixinClasses("mcpatcherforge.ctm_cc.MixinTextureMap")
     ),
     //End from NotFine
-
+//SkidZoom
+    SKIDZOOM(new Builder("zoom zoom")
```
Comment (Alexdoru) [inline src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java:534]: please remove all the word "skid" that you put everywhere, it means "stealing code" and that is not appripriate

File: src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java [hunk 5]
```diff
@@ -530,7 +530,13 @@ public enum Mixins {
...
+        .setSide(Side.CLIENT).setPhase(Phase.LATE)
```
Comment (Alexdoru) [inline src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java:535]: if it's vanilla it should be Phase.EARLY

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+import net.minecraft.client.settings.KeyBinding;
+import net.minecraftforge.client.event.FOVUpdateEvent;
+import net.minecraftforge.common.MinecraftForge;
+import org.lwjgl.input.Keyboard;
+import org.lwjgl.input.Mouse;
+
+@Mod(modid = "skidzoom", name = "SkidZoom", version = "1.0")
+public class SkidZoom {
+
+    public KeyBinding zoomKey;
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:16]: doesn't need to be public

File: src/mixin/java/com/gtnewhorizons/angelica/mixins/late/client/skidzoom/InventoryPlayerMixin.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import net.loudcats.wyatt.skidzoom.SkidZoom;
+import net.minecraft.entity.player.InventoryPlayer;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(InventoryPlayer.class)
+public abstract class InventoryPlayerMixin {
+    @Inject(method = "changeCurrentItem", at = @At("HEAD"), cancellable = true)
```
Comment (Alexdoru) [inline src/mixin/java/com/gtnewhorizons/angelica/mixins/late/client/skidzoom/InventoryPlayerMixin.java:12]: I would prefer that you wrap the `int i = Mouse.getEventDWheel();` line in the `Minecraft.runTick()` method and make it return 0 when it's zooming

**Review by Alexdoru — CHANGES_REQUESTED** [review]

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    private static boolean zoom;
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:17]: why is one field static and not the 2 others

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    public void init(FMLInitializationEvent event) {
+        MinecraftForge.EVENT_BUS.register(this);
+        zoomKey = new KeyBinding("key.skidzoom", Keyboard.KEY_C, "key.categories.misc");
+        zoom = false;
+        ClientRegistry.registerKeyBinding(zoomKey);
+    }
+
+    @SubscribeEvent
+    public void ZoomyZoomZoom(FOVUpdateEvent e) {
+        if (zoomKey.getIsKeyPressed() || zoomKey.isPressed()) {
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:30]: don't use `isPressed()`, only use `getIsKeyPressed()`

File: src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            zoom = true;
+            if (zoom) {
```
Comment (Alexdoru) [inline src/main/java/net/loudcats/wyatt/skidzoom/SkidZoom.java:32]: `zoom = true` followed by `if(zoom)` is this thing made by chat GPT ? and then you make us review it ?

**Review by Alexdoru — CHANGES_REQUESTED** [review]

**Review by Alexdoru — CHANGES_REQUESTED** [review]
also your implementation is broken, it will stayed zoomed for a bit if you unpress the key because the even you hook into isn't fired for every frame. You need to mixin into the rendering loop and modify the FOV in there

Also you should add a setting to have a "smooth zoom" and "instant zoom"

**Comment by Alexdoru** [comment]
I'm gonna do my own

File: src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java (hunk 4 above)
Comment (WyattTheSkid) [reply src/main/java/com/gtnewhorizons/angelica/mixins/Mixins.java:534]: I'm sorry I just seen all these. I'm not super familiar with the way that pull requests work and stuff so I missed all of these my apologies. And as for the word skid, its a copy pasta implementation of a mod I made for myself a while ago that was also called SkidZoom so I apologize for not seeing these messages, that is all. I will turn on notifications for any potential future pull requests.

**Pull request closed without merging**